src/request.rs

```diff
@@ -238,27 +267,176 @@ impl RequestBuilder {
+    ///
+    /// The headers will be merged in to any already set.
+    pub fn headers(&mut self, headers: ::header::Headers) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().headers.extend(headers.iter());
+        self
+    }
+
+    /// Enable HTTP basic authentication.
+    pub fn basic_auth<U, P>(&mut self, username: U, password: Option<P>) -> &mut MultipartRequestBuilder
+        where U: Into<String>,
+              P: Into<String>
+    {
+        self.request_builder.header(::header::Authorization(::header::Basic {
+                                                username: username.into(),
+                                                password: password.map(|p| p.into()),
+                                            }));
+        self
+    }
+
+    /// Set the request body.
+    pub fn body<T: Into<Body>>(&mut self, body: T) -> &mut MultipartRequestBuilder {
+        self.request_builder.request_mut().body = Some(body.into());
+        self
+    }
+
+    /// Add parameters to the multipart/form-data `Request`.
+    pub fn params(&mut self, params: Params) -> &mut MultipartRequestBuilder {
+        self.params = Some(params);
+        self
+    }
+
+    /// Add files to the multipart/form-data `Request`.
+    pub fn files(&mut self, files: Files) -> &mut MultipartRequestBuilder {
+        self.files = Some(files);
+        self
+    }
+
+    /// Build a `Request`, which can be inspected, modified and executed with
+    /// `Client::execute()`.
+    ///
+    /// # Panics
+    ///
+    /// This method consumes builder internal state. It panics on an attempt to
+    /// reuse already consumed builder.
+    pub fn build(&mut self) -> ::Result<Request> {
+        let mut body: Vec<u8> = Vec::new();
+        let boundary = MultipartRequestBuilder::choose_boundary();
+        let multipart_mime = ContentType(format!{"multipart/form-data; boundary={}", boundary}
+                                             .parse::<mime::Mime>()
+                                             .unwrap());
+        if let Some(p) = self.params.clone() {
+            for (name, value) in p {
+                write_bytes!(body, "\r\n--{}\r\n", boundary);
+                write_bytes!(body, "Content-Disposition: form-data; name=\"{}\"\n", name);
+                write_bytes!(body, "\r\n{}", value);
+            }
+        }
+
+        if let Some(f) = self.files.clone() {
+            for File { name, path, mime } in f {
+                write_bytes!(body, "\r\n--{}\r\n", boundary);
+                write_bytes!(body, "Content-Disposition: form-data; name=\"{}\"", name);
+                write_bytes!(body,
+                             "; filename=\"{}\"",
+                             path.file_name().unwrap().to_str().unwrap());
+                write_bytes!(body, "\r\nContent-type: {}\r\n\r\n", mime);
+                let mut content = try_!(fs::File::open(path));
+                content.read_to_end(&mut body).unwrap();
```

Comment:
The files could be very big, so instead of copying all of them into memory, it'd probably be better to create an internal `MultipartBody` that implements `Read`, and pushes the multipart bytes onto the read buf, and then forwards to `File::read`.